src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+                    .worker_terminated(child_runtime_name.clone())
+                    .await;
+                children_specs.push(failed_child_spec);
+            }
+        }
+    }
+    if termination_errors.is_empty() {
+        Ok(children_specs)
+    } else {
+        Err((children_specs, TerminationError(termination_errors)))
+    }
+}
+
+impl Spec {
+    pub fn new(
+        name: impl Into<String>,
+        children: Vec<worker::Spec>,
+        ev_notifier: EventNotifier,
+    ) -> Spec {
+        Spec {
+            name: name.into(),
+            ev_notifier,
+            children,
+
+            order: Order::LeftToRight,
+            strategy: Strategy::OneForOne,
+        }
+    }
+
+    async fn start_supervisor<'a>(
+        self,
+        parent_ctx: &context::Context,
+        parent_name: &'a str,
+    ) -> Result<Supervisor, (Spec, Arc<StartError>)> {
+        let (ctx, _terminate_supervisor) = Context::with_cancel(parent_ctx);
+
+        let name = self.name.clone();
+        let mut ev_notifier = self.ev_notifier.clone();
+        let order = self.order.clone();
+        let strategy = self.strategy.clone();
+
+        let sup_runtime_name = format!("{}/{}", parent_name, name);
+        let result = start_children(
+            &ctx,
+            ev_notifier.clone(),
+            &sup_runtime_name,
+            &order,
+            self.children,
+        )
+        .await;
+
+        match result {
+            Err((children_specs, err0)) => {
+                let err = Arc::new(err0);
+                ev_notifier
+                    .supervisor_start_failed(&sup_runtime_name, err.clone())
+                    .await;
+                let spec = Spec {
+                    name,
+                    ev_notifier,
+                    order,
+                    strategy,
+                    children: children_specs,
+                };
+                return Err((spec, err));
+            }
+            Ok(children) => {
+                ev_notifier.supervisor_started(&sup_runtime_name).await;
+                Ok(Supervisor {
+                    runtime_name: sup_runtime_name.to_owned(),
+                    children,
+
+                    name,
+                    ev_notifier,
+                    order,
+                    strategy,
+                })
+            }
+        }
+    }
+
+    pub async fn start(
+        self,
+        parent_ctx: &context::Context,
+    ) -> Result<Supervisor, (Spec, Arc<StartError>)> {
+        let root_name = "";
+        self.start_supervisor(parent_ctx, root_name).await
+    }
+}
+
+impl Supervisor {
+    pub async fn terminate(self) -> Result<Spec, (Spec, Arc<TerminationError>)> {
+        let name = self.name;
+        let runtime_name = self.runtime_name;
+        let order = self.order;
+        let strategy = self.strategy;
+        let mut ev_notifier = self.ev_notifier;
+        let runtime_children = ordered_children_on_termination(&order, self.children);
```

Comment:
You can destructure a struct in a `let` binding just like an enum.

```suggestion
        let Supervisor { 
            name, 
            runtime_name, 
            order, strategy, 
            mut ev_notifier, 
            children 
        } = self;
        let runtime_children = ordered_children_on_termination(&order, children);
```